### ISSUE TITLE
Add Positional Evalutions for Heuristic

### ISSUE THREAD

**Issue by manumei** [manudelp/ultimate-tic-tac-toe#69]
# Purpose
It's important to not forget why I'm adding and modifying all these things in the heuristic function, and to keep in mind what functionality everything I'm doing actually serves, cause any last-minute realizations or changes may also adversely affect the importance or purpose of the other changes. 

### Board Evaluation
Serves for the general balance by evaluating each individual local board with actual heuristic values (line evaluations), contributes to the Global Board Balance both in the general sum of all local evaluations, and as main weight in the calculation of the "Positional Score" that is then used for the Global Balance as well.

### Board Result
Used for determining the connectivity of won boards and thus see a difference between 2 won boards completely apart, or connected to potentially crown a third, without the need of reaching won-game depth.

### Positional Lead
Is used to determine the connectivity of neighbors for each player, since neighborhood hash utilizes the values of Positional Lead Values (-3, 0, 4, 3) to determine what the neighborhood value of each local board is, for the Positional Score to be implemented. 

### Positional Score
Combines Evaluation & Best Connection to determine the positional advantage of each local board for its respective lead player, to then be used in the Balance for some insight on how a player can set up future connected results instead of playing unconnected.

# Instructions
Granted, some may be a bit outdated considering some of my last-minute realizations or new ideas

## Step 1
Modify Glob Eval Hash to return two extra elements, given now as
{board : (evaluation, result, positional lead, positional score)}
With the following meanings:

**Evaluation (float)** ✅
Already done! Local Evaluation of the board given the sum of Line Evals (prone to be edited for enhancement)

**Result (int)** ✅
Already done! Current result of the local board as ±1 for a winner, 2 for a draw, 0 for on-going.

**Positional Lead (int)** ☑️
Value representing who currently has a better balance and leading position in the board, comparing the local board evaluations. The Positional Lead can be either ±3 for each respective player, or 0 if they are currently tied, and 4 if the board is a full tie.

**Positional Score (int)** ☑️
If result != 0, then position is irrelevant, and [Positional Score = 0]
Else, if the board is still on-going, [Positional Score = Positional Lead/3 × Best Connection]
_Maybe multiply positional score by *= e^((x^2)/200)), where x is the local board eval value_

**MAKE SURE THAT POSITIONAL SCORE REWARD REMAINS LOWER THAN RESULTS SCORE REWARD!**

**Auxiliaries needed**

Best Connection (int >= 0) ☑️
Represents the best connection of the player that is currently winning, or 0 if they are currently tied. If the best connection is a single-piece-line, then Best Connection = 2.4, if it's a double-piece-line, then Best Connection = 5. Must always be positive, the sign is already differentiated in Positional Lead value.
• If Best Connection is a triple-piece, raise ValueError since Best Connection is only for when there is no winner
• If Best Connection is a zero-piece, raise ValueError since Best Connection is only for when there is no draw

Some less-obvious debugging ideas when adding this:
- if ((Evaluation != 0 and Positional_Score == 0) or (Evaluation == 0 and Positional_Score != 0)): raise ValueError
- if (Result != 0 and Positional_Lead != 4): raise ValueError
- if (Result != 0 and Positional_Score != 0): raise ValueError
- if ((Positional_Lead == 0 or Best_Connection == 0) and (Positional_Lead != Best_Connection)): raise ValueError

Create and run the necessary HRTs to test this new hash feature of course.

## Step 2
Create a 2D numpy array of size 9 with the Positional Leads. And pass it through the get_positional_coef_hash() function that given the size 9 array, will itself return the return a size 9 tuple, with the positional coefficients of each local board. Positional Coefficient is calculated given their neighbors and unblocked lines of the currently evaluated local board. 
If all its lines are blocked, the coefficient will be 0, otherwise it will be:
[single_coef × (amount of open empty lines where it's alone) + double_coef × (amount of open lines where it connects for double)]
Initial placeholder values could be single_coef = 0.12, double_coef = 0.52.
Lines are not counted as open for a player if (line.count(4) > 0) or (line.count(-player) > 0). If so, their multiplier is 0

Then, once you obtain all the coefficients, you will calculate the Positional Results of each local board, by performing the operation
∑ [Positional_Score_ij × Positional_Coefficient_ij] for each board (i, j). And add this result to the final balance in Board Balance.

For this you will need to load up a new hash of 4**9 almost equal to the results board hash, but now using 3s, -3s, 4s and 0s.
And also, different in how it evaluates. This one returns a Tuple of 9 elements, each i%3+j element representing the (i, j) local board's coefficient.
